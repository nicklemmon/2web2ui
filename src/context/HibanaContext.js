import React, { createContext, useContext } from 'react';
import { connect } from 'react-redux';
import { updateUserUIOptions } from 'src/actions/currentUser';
import { showAlert } from 'src/actions/globalAlert';
import { isUserUiOptionSet } from 'src/helpers/conditions/user';
import { selectCondition } from 'src/selectors/accessConditionState';
import { segmentTrack, SEGMENT_EVENTS } from '../helpers/segment';

export const HibanaStateContext = createContext();

function Provider(props) {
  const { children, ...rest } = props;

  return <HibanaStateContext.Provider value={rest}>{children}</HibanaStateContext.Provider>;
}

function mapStateToProps(state) {
  return {
    isCurrentUserPending: state.currentUser.loading,
    isHibanaEnabled: selectCondition(isUserUiOptionSet('isHibanaEnabled'))(state),
    isBannerVisible: selectCondition(isUserUiOptionSet('isHibanaBannerVisible'))(state),
  };
}

const mapDispatchToProps = {
  dismissBanner: () => updateUserUIOptions({ isHibanaBannerVisible: false }),
  setIsHibanaEnabled: bool => {
    if (window.pendo && window.pendo.track) {
      window.pendo.track(`Hibana Toggle - ${Boolean(bool) ? 'On' : 'Off'}`);
    }
    if (bool) {
      segmentTrack(SEGMENT_EVENTS.HIBANA_TOGGLED_ON);
    } else {
      segmentTrack(SEGMENT_EVENTS.HIBANA_TOGGLED_OFF);
    }

    // Always dismiss the banner when the user toggles their theme
    return updateUserUIOptions({ isHibanaEnabled: bool, isHibanaBannerVisible: false });
  },
  showAlert,
};

export const HibanaProvider = connect(mapStateToProps, mapDispatchToProps)(Provider);

export function useHibana() {
  const context = useContext(HibanaStateContext);
  if (context === undefined) throw new Error('useHibana must be used within a HibanaProvider');

  return [context];
}
