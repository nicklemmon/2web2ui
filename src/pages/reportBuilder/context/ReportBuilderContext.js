import React, { useCallback, useContext, useMemo, useReducer, createContext } from 'react';
import moment from 'moment';
import { connect } from 'react-redux';
import { selectFeatureFlaggedMetrics } from 'src/selectors/metrics';
import { getRelativeDates } from 'src/helpers/date';
import {
  getMetricsFromKeys,
  getPrecision as getRawPrecision,
  getRollupPrecision,
} from 'src/helpers/metrics';
import { getLocalTimezone } from 'src/helpers/date';
import { stringifyTypeaheadfilter } from 'src/helpers/string';
import config from 'src/config';
import { FILTER_KEY_MAP } from 'src/helpers/metrics';
import {
  getIterableFormattedGroupings,
  getApiFormattedGroupings,
  hydrateFilters,
} from '../helpers';

const ReportOptionsContext = createContext({});

const initialState = {
  filters: [],
  comparisons: [],
};

const reducer = (state, action) => {
  switch (action.type) {
    case 'ADD_FILTERS': {
      return {
        ...state,
        filters: [
          ...state.filters,
          { AND: { [FILTER_KEY_MAP[action.payload.type]]: { eq: [action.payload] } } },
        ],
      };
    }
    case 'UPDATE_REPORT_OPTIONS': {
      const { payload, meta } = action;
      const { useMetricsRollup, subaccounts } = meta;
      let update = { ...state, ...payload };
      const getPrecision = useMetricsRollup ? getRollupPrecision : getRawPrecision;

      if (!update.timezone || !useMetricsRollup) {
        update.timezone = getLocalTimezone();
      }

      if (!update.metrics) {
        update.metrics = config.reportBuilder.defaultMetrics;
      }

      if (payload.filters) {
        update.filters = hydrateFilters(payload.filters, { subaccounts });
      }

      if (!update.relativeRange) {
        update.relativeRange = '7days';
      }

      //old version of update

      const rollupPrecision = useMetricsRollup && (update.precision || 'hour'); //Default to hour since it's the recommended rollup precision for 7 days
      if (update.relativeRange !== 'custom') {
        const { from, to } = getRelativeDates(update.relativeRange, {
          precision: rollupPrecision,
        });
        //for metrics rollup, when using the relative dates, get the precision, else use the given precision
        //If precision is not in the URL, get the recommended precision.
        const precision = getPrecision({ from, to, precision: rollupPrecision });
        update = { ...update, from, to, precision };
      } else {
        const precision = getPrecision({
          from: update.from,
          to: moment(update.to),
          precision: rollupPrecision,
        });

        update = { ...update, precision };
      }
      return {
        ...update,
        isReady: true,
      };
    }

    case 'REMOVE_FILTER': {
      const { payload } = action;
      const groupings = getIterableFormattedGroupings(state.filters);
      const targetFilter = groupings[payload.groupingIndex].filters[payload.filterIndex];
      // Filter out matching values based on index
      targetFilter.values = targetFilter.values.filter(
        value => value !== targetFilter.values[payload.valueIndex],
      );
      // Remap iterable data structure to match API structure
      const filters = getApiFormattedGroupings(groupings);

      return {
        ...state,
        filters,
      };
    }

    default:
      throw new Error(`${action.type} is not supported.`);
  }
};

const getSelectors = reportOptions => {
  const selectDateOptions = {
    from: moment(reportOptions.from)
      .utc()
      .format(),
    to: moment(reportOptions.to)
      .utc()
      .format(),
    range: reportOptions.relativeRange,
    timezone: reportOptions.timezone,
    precision: reportOptions.precision,
  };

  const selectTypeaheadFilters = {
    filters: (reportOptions.filters || []).map(stringifyTypeaheadfilter),
  };

  const selectCompareFilters = {
    comparisons: reportOptions.comparisons || [],
  };

  const selectSummaryMetrics = {
    metrics: (reportOptions.metrics || []).map(metric =>
      typeof metric === 'string' ? metric : metric.key,
    ),
    //TODO RB CLEANUP: can probably remove the check if it's an object
  };

  const selectSummaryMetricsProcessed = getMetricsFromKeys(reportOptions.metrics || [], true);

  /**
   * Converts reportOptions for url sharing
   */
  const selectReportSearchOptions = { ...selectDateOptions, ...selectTypeaheadFilters };

  /**
   * Converts reportOptions for url sharing for the summary chart
   */
  const selectSummaryChartSearchOptions = {
    ...selectDateOptions,
    ...selectTypeaheadFilters,
    ...selectSummaryMetrics,
    ...selectCompareFilters,
  };

  return {
    selectSummaryMetricsProcessed,
    selectReportSearchOptions,
    selectSummaryChartSearchOptions,
  };
};

const ReportOptionsContextProvider = props => {
  const { useMetricsRollup, subaccounts } = props;
  const [state, dispatch] = useReducer(reducer, initialState);

  const refreshReportOptions = useCallback(
    payload => {
      return dispatch({
        type: 'UPDATE_REPORT_OPTIONS',
        payload,
        meta: { useMetricsRollup, subaccounts },
      });
    },
    [dispatch, useMetricsRollup, subaccounts],
  );

  const addFilters = useCallback(
    payload => {
      return dispatch({
        type: 'ADD_FILTERS',
        payload,
      });
    },
    [dispatch],
  );

  const removeFilter = useCallback(
    payload => {
      return dispatch({
        type: 'REMOVE_FILTER',
        payload,
      });
    },
    [dispatch],
  );

  //Not currently used but I am leaving them here for now.
  const setFilters = useCallback(
    payload => {
      return dispatch({
        type: 'SET_FILTERS',
        payload,
      });
    },
    [dispatch],
  );

  //Not currently used but I am leaving them here for now.
  const clearFilters = useCallback(() => {
    return dispatch({
      type: 'CLEAR_FILTERS',
    });
  }, [dispatch]);

  const actions = {
    addFilters,
    clearFilters,
    setFilters,
    removeFilter,
    refreshReportOptions,
  };
  const selectors = useMemo(() => getSelectors(state), [state]);

  return (
    <ReportOptionsContext.Provider value={{ state, actions, selectors }}>
      {props.children}
    </ReportOptionsContext.Provider>
  );
};

const mapStateToProps = state => ({
  subaccounts: state.subaccounts.list,
  useMetricsRollup: selectFeatureFlaggedMetrics(state).useMetricsRollup,
});

export const ReportBuilderContextProvider = connect(mapStateToProps)(ReportOptionsContextProvider);

export const useReportBuilderContext = () => useContext(ReportOptionsContext);
