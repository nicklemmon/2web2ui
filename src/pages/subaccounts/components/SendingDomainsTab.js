import React, { Component } from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { PageLink } from 'src/components/links';
import { Button, Panel, Stack } from 'src/components/matchbox';
import { Loading } from 'src/components';
import { TableCollection, DomainStatusCell, StatusTooltipHeader } from 'src/components';
import { selectSendingDomainsForSubaccount } from 'src/selectors/sendingDomains';
import { HibanaStateContext } from 'src/context/HibanaContext';

const StyledPanelContent = styled.div`
  text-align: center;
`;

const columns = [
  { label: 'Domain', width: '30%', sortKey: 'domain' },
  { label: <StatusTooltipHeader />, width: '40%' },
];

export const getRowData = (row, isHibanaEnabled) => [
  <PageLink
    to={
      isHibanaEnabled
        ? `/domains/details/sending-bounce/${row.domain}`
        : `/account/sending-domains/edit/${row.domain}`
    }
  >
    {row.domain}
  </PageLink>,
  <DomainStatusCell domain={row} />,
];

const getToLink = (value = {}) =>
  value.isHibanaEnabled ? '/domains/list/sending' : '/account/sending-domains';

export class SendingDomainsTab extends Component {
  render() {
    const { loading } = this.props;

    if (loading) {
      return <Loading minHeight="300px" />;
    }

    const showEmpty = this.props.domains.length === 0;
    return (
      <HibanaStateContext.Consumer>
        {value => (
          <>
            {showEmpty ? (
              <Panel.LEGACY>
                <Panel.LEGACY.Section>
                  <StyledPanelContent>
                    <Stack>
                      <p>
                        This subaccount has no sending domains assigned to it. You can assign an
                        existing one, or create a new one.
                      </p>
                      <div>
                        <PageLink as={Button} variant="secondary" to={getToLink(value)}>
                          Manage Sending Domains
                        </PageLink>
                      </div>
                    </Stack>
                  </StyledPanelContent>
                </Panel.LEGACY.Section>
              </Panel.LEGACY>
            ) : (
              <>
                <Panel.LEGACY marginBottom="0" borderBottom="0">
                  <Panel.LEGACY.Section>
                    <p>Sending Domains assigned to this subaccount.</p>
                  </Panel.LEGACY.Section>
                </Panel.LEGACY>
                <TableCollection
                  columns={columns}
                  getRowData={row => getRowData(row, value.isHibanaEnabled)}
                  pagination={true}
                  rows={this.props.domains}
                />{' '}
              </>
            )}
          </>
        )}
      </HibanaStateContext.Consumer>
    );
  }
}

const mapStateToProps = (state, props) => ({
  loading: state.sendingDomains.listLoading,
  domains: selectSendingDomainsForSubaccount(state, props),
});

export default withRouter(connect(mapStateToProps)(SendingDomainsTab));
