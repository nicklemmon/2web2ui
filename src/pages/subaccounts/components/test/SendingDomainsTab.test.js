import React from 'react';
import { shallow } from 'enzyme';
import { SendingDomainsTab, getRowData } from '../SendingDomainsTab';
import { HibanaStateContext } from 'src/context/HibanaContext';

const mockConsumer = jest.fn();

describe('SendingDomainsTab', () => {
  beforeEach(() => {
    mockConsumer.mockReset();
  });

  const defaultProps = {
    domains: [
      {
        domain: 'foo.com',
      },
      {
        domain: 'bar.com',
      },
    ],
    loading: false,
  };

  const subject = props => shallow(<SendingDomainsTab {...defaultProps} {...props} />);

  it('should load domains in tab', () => {
    mockConsumer.mockReturnValue({ isHibanaEnabled: false });
    const wrapper = subject()
      .find(HibanaStateContext.Consumer)
      .renderProp('children')(mockConsumer());
    expect(wrapper).toHaveTextContent('Sending Domains assigned to this subaccount.');
    expect(wrapper.find('TableCollection')).toHaveProp('rows', defaultProps.domains);
  });

  it('should show panel loading while loading domains', () => {
    mockConsumer.mockReturnValue({ isHibanaEnabled: false });
    const wrapper = subject({ loading: true });
    expect(wrapper.find('Loading')).toExist();
    expect(wrapper.find('TableCollection')).not.toExist();
  });

  it('should show empty message when 0 domains exist', () => {
    mockConsumer.mockReturnValue({ isHibanaEnabled: false });
    const wrapper = subject({ domains: [] })
      .find(HibanaStateContext.Consumer)
      .renderProp('children')(mockConsumer());

    expect(wrapper).toHaveTextContent(
      'This subaccount has no sending domains assigned to it. You can assign an existing one, or create a new one.',
    );
    expect(wrapper.find('PageLink')).toHaveProp('to', '/account/sending-domains');
  });

  it('should show a link to /domains/list/sending when hibana is enabled', () => {
    mockConsumer.mockReturnValue({ isHibanaEnabled: true });
    const wrapper = subject({ domains: [] }).renderProp('children')(mockConsumer());
    expect(wrapper.find('PageLink')).toHaveProp('to', '/domains/list/sending');
  });

  it('snapshots getRowData without hibana', () => {
    expect(getRowData({ domain: 'foo.com' }, false)).toMatchSnapshot();
  });

  it('snapshots getRowData with hibana', () => {
    expect(getRowData({ domain: 'foo.com' }, true)).toMatchSnapshot();
  });
});
