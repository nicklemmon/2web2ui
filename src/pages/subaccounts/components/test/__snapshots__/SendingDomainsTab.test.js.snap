// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendingDomainsTab snapshots getRowData with hibana 1`] = `
Array [
  <PageLink
    to="/domains/details/sending-bounce/foo.com"
  >
    foo.com
  </PageLink>,
  <DomainStatusCell
    domain={
      Object {
        "domain": "foo.com",
      }
    }
  />,
]
`;

exports[`SendingDomainsTab snapshots getRowData without hibana 1`] = `
Array [
  <PageLink
    to="/account/sending-domains/edit/foo.com"
  >
    foo.com
  </PageLink>,
  <DomainStatusCell
    domain={
      Object {
        "domain": "foo.com",
      }
    }
  />,
]
`;
